File: schemakeeper-server/src/test/scala/schemakeeper/server/avro/AvroSchemaCompatibilityTest.scala
package schemakeeper.server.avro

import org.apache.avro.{Schema, SchemaBuilder}
import org.scalatest.{Matchers, WordSpec}

class AvroSchemaCompatibilityTest extends WordSpec with Matchers {
  "NONE compatibility validator" should {
    val validator = AvroSchemaCompatibility.NONE_VALIDATOR

    "always return true" in {
      val a = Schema.create(Schema.Type.INT)
      val b = Schema.create(Schema.Type.STRING)
      val c = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .optionalString("f1")
        .requiredInt("f2")
        .optionalDouble("f3")
        .nullableString("f5", "default")
        .endRecord()

      assert(validator.isCompatible(a, b))
      assert(validator.isCompatible(b, c))
      assert(validator.isCompatible(a, c))
    }
  }

  "BACKWARD compatibility validator" should {
    val validator = AvroSchemaCompatibility.BACKWARD_VALIDATOR

    "return true - add new optional field and remove one required and one optional" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .optionalString("f3")
        .endRecord()

      assert(validator.isCompatible(newSchema, previousSchema))
    }

    "return false - add new required field" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .requiredString("f3")
        .endRecord()

      assert(!validator.isCompatible(newSchema, previousSchema))
    }

    "return true - nested field type" in {
      val firstSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val secondSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .nullableString("f2", "default")
        .endRecord()
      val thirdSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .nullableString("f2", "default")
        .name("nested")
          .`type`()
          .optional()
          .record("nested")
          .fields()
          .optionalString("nested_f1")
          .endRecord()
        .endRecord()

      assert(validator.isCompatible(secondSchema, firstSchema))
      assert(validator.isCompatible(thirdSchema, secondSchema))
    }
  }

  "FORWARD compatibility validator" should {
    val validator = AvroSchemaCompatibility.FORWARD_VALIDATOR

    "return true - delete optional field and add required and optional fields" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .requiredString("f3")
        .optionalString("f4")
        .endRecord()

      assert(validator.isCompatible(newSchema, previousSchema))
    }

    "return false - delete required field" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .optionalString("f2")
        .endRecord()

      assert(!validator.isCompatible(newSchema, previousSchema))
    }
  }

  "FULL compatibility validator - add optional field" should {
    val validator = AvroSchemaCompatibility.FULL_VALIDATOR

    "return true" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .optionalString("f3")
        .endRecord()

      assert(validator.isCompatible(newSchema, previousSchema))
    }

    "return false - delete required field" in {
      val previousSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .optionalString("f2")
        .endRecord()

      assert(!validator.isCompatible(newSchema, previousSchema))
    }
  }

  "BACKWARD TRANSITIVE compatibility validator" should {
    val validator = AvroSchemaCompatibility.BACKWARD_TRANSITIVE_VALIDATOR

    "return true - add optional fields" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .optionalString("f3")
        .endRecord()

      assert(validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }

    "return false - add required field" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .optionalString("f3")
        .requiredString("f4")
        .endRecord()

      assert(!validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }
  }

  "FORWARD TRANSITIVE compatibility validator" should {
    val validator = AvroSchemaCompatibility.FORWARD_TRANSITIVE_VALIDATOR

    "return true - add required and remove optional fields" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .requiredString("f3")
        .endRecord()

      assert(validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }

    "return false - delete required" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f3")
        .endRecord()

      assert(!validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }
  }

  "FULL TRANSITIVE compatibility validator" should {
    val validator = AvroSchemaCompatibility.FULL_TRANSITIVE_VALIDATOR

    "return true - add optional fields" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .optionalString("f3")
        .endRecord()

      assert(validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }

    "return false - add required field" in {
      val previousSchema1 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .endRecord()
      val previousSchema2 = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .endRecord()
      val newSchema = SchemaBuilder
        .builder("namespace")
        .record("record")
        .fields()
        .requiredString("f1")
        .optionalString("f2")
        .requiredString("f3")
        .endRecord()

      assert(!validator.isCompatible(newSchema, Seq(previousSchema1, previousSchema2)))
    }
  }
}

File: modules/common/src/test/java/schemakeeper/schema/AvroSchemaCompatibilityTest.java
package schemakeeper.schema;

import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;
import org.junit.Test;

import java.util.Arrays;
import static org.junit.Assert.*;


public class AvroSchemaCompatibilityTest {
    @Test
    public void nullPreviousSchemaShouldReturnTrueForAllCompatibilityTypes() {
        Schema a = Schema.create(Schema.Type.INT);

        assertTrue(AvroSchemaCompatibility.NONE_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.BACKWARD_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.FORWARD_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.FULL_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.BACKWARD_TRANSITIVE_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.FORWARD_TRANSITIVE_VALIDATOR.isCompatible(a, (Schema) null));
        assertTrue(AvroSchemaCompatibility.FULL_TRANSITIVE_VALIDATOR.isCompatible(a, (Schema) null));
    }

    @Test
    public void noneCompatibilityValidatorAlwaysReturnTrue() {
        Schema a = Schema.create(Schema.Type.INT);
        Schema b = Schema.create(Schema.Type.STRING);
        Schema c = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .requiredString("f2")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.NONE_VALIDATOR.isCompatible(a, b));
        assertTrue(AvroSchemaCompatibility.NONE_VALIDATOR.isCompatible(b, c));
        assertTrue(AvroSchemaCompatibility.NONE_VALIDATOR.isCompatible(a, c));
    }

    @Test
    public void backwardCompatibilityReturnTrue() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // add optional field
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .optionalString("f3")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.BACKWARD_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void backwardCompatibilityReturnFalse() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // add required field
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .requiredString("f3")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.BACKWARD_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void backwardCompatibilityReturnTrueWithNestedFields() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .optionalString("f3")
                .endRecord();

        Schema schemaWithNestedField = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .optionalString("f3")
                .name("nested")
                .type()
                .optional()
                .record("nested")
                .fields()
                .optionalString("nested_f1")
                .endRecord()
                .endRecord();

        assertTrue(AvroSchemaCompatibility.BACKWARD_VALIDATOR.isCompatible(newSchema, previousSchema));
        assertTrue(AvroSchemaCompatibility.BACKWARD_VALIDATOR.isCompatible(schemaWithNestedField, newSchema));
    }

    @Test
    public void forwardCompatibilityReturnTrue() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // remove optional and add required and optional fields
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .requiredString("f3")
                .optionalString("f4")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.FORWARD_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void forwardCompatibilityReturnFalse() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // delete required field
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .optionalString("f3")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.FORWARD_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void fullCompatibilityReturnTrue() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // remove optional and add fields
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f3")
                .optionalString("f4")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.FULL_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void fullCompatibilityReturnFalse() {
        Schema previousSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();

        // delete required field
        Schema newSchema = SchemaBuilder
                .builder()
                .record("test")
                .fields()
                .optionalString("f2")
                .optionalString("f3")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.FULL_VALIDATOR.isCompatible(newSchema, previousSchema));
    }

    @Test
    public void backwardTransitiveCompatibilityReturnTrue() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .optionalString("f3")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.BACKWARD_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }

    @Test
    public void backwardTransitiveCompatibilityReturnFalse() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .optionalString("f3")
                .requiredString("f4")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.BACKWARD_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }

    @Test
    public void forwardTransitiveCompatibilityReturnTrue() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .requiredString("f3")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.FORWARD_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }

    @Test
    public void forwardTransitiveCompatibilityReturnFalse() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f3")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.FORWARD_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }

    @Test
    public void fullTransitiveCompatibilityReturnTrue() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .optionalString("f3")
                .endRecord();

        assertTrue(AvroSchemaCompatibility.FULL_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }

    @Test
    public void fullTransitiveCompatibilityReturnFalse() {
        Schema previousSchema1 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .endRecord();
        Schema previousSchema2 = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .endRecord();
        Schema newSchema = SchemaBuilder
                .builder("namespace")
                .record("record")
                .fields()
                .requiredString("f1")
                .optionalString("f2")
                .requiredString("f3")
                .endRecord();

        assertFalse(AvroSchemaCompatibility.FULL_TRANSITIVE_VALIDATOR.isCompatible(newSchema, Arrays.asList(previousSchema1, previousSchema2)));
    }
}